sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java

```diff
@@ -371,14 +371,27 @@ public void replaceChild(Node existing, Node replacement) {
      * extracts a component of a tuple, or other operations that occur at pipeline assembly time.
      */
     public boolean isCompositeNode() {
-      return !parts.isEmpty() || isRootNode() || returnsOthersOutput();
+      if (isRootNode()) {
+        LOG.info("Composite because root node {}", this);
+        return true;
+      }
+      if (!parts.isEmpty()) {
+        LOG.info("Composite because has subtransforms {}", this);
+        return true;
+      }
+      if (returnsOthersOutput()) {
+        LOG.info("Composite because returns others output {}", this);
```

Comment:
Yea, it can. This is why `CreateDataflowView.forBatch` was considered a composite. Because it materializes its input and then returns its input. So it returns another transform's output and the traversal thinks it is a composite. That is why I had to add the extra `materializatoinInput = input.apply(<identity>)` bit.